pkg/svcproxy/svcproxy.go

```diff
@@ -0,0 +1,154 @@
+package svcproxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/netip"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"github.com/rs/zerolog"
+
+	"encr.dev/pkg/logging"
+)
+
+type SvcProxy struct {
+	listener   net.Listener
+	logger     zerolog.Logger
+	httpServer *http.Server
+
+	mu       sync.RWMutex
+	gateways map[string]*httputil.ReverseProxy // Map of the gateway name to address and port it's listening on
+	services map[string]*httputil.ReverseProxy // Map of service name to address and port it's listening on
+}
+
+var (
+	_ http.Handler = (*SvcProxy)(nil)
+)
+
+// New creates a new service proxy and it starts listening on a random port
+//
+// You must call Close() on the returned service proxy when you are done with it.
+func New(ctx context.Context, logger zerolog.Logger) (*SvcProxy, error) {
+	var lc net.ListenConfig
+	ln, err := lc.Listen(ctx, "tcp", "127.0.0.1:0")
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to listen")
+	}
+
+	proxy := &SvcProxy{
+		listener: ln,
+		logger:   logger,
+		gateways: make(map[string]*httputil.ReverseProxy),
+		services: make(map[string]*httputil.ReverseProxy),
+	}
+
+	proxy.httpServer = &http.Server{
+		Addr:        ln.Addr().String(),
+		BaseContext: func(_ net.Listener) context.Context { return ctx },
+		Handler:     proxy,
+	}
+
+	go func() {
+		if err := proxy.httpServer.Serve(ln); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			logger.Err(err).Msg("error serving")
+		}
+	}()
+
+	return proxy, nil
+}
+
+// Close stops running the service proxy
+func (p *SvcProxy) Close() {
+	if p.httpServer != nil {
```

Comment:
When would this be nil?

Reply:
in an earlier iteration there was a path when it could have been, gone now